### ISSUE TITLE
Feature request: Live view proxy

### ISSUE THREAD

**Issue by anthonyangel** [tchellomello/python-ring-doorbell#23]
Hi, would it be possible to use this library to allow a continuous live view stream to be exposed locally as an IP camera feed (either ONVIF or just an mp4 stream), so that it could be recorded by an NVR?

Ideally this would bypass Rings servers and just stay within the LAN (would reduce the overheads in terms of data transmission), but that doesn't seem possible at the moment,  but this would allow Ring to be integrated into an existing CCTV installation instead of existing as a separate system.

Thanks,
Anthony

**Comment by tchellomello** [comment]
@anthonyangel  if you call the function `live_streaming_json` you will get a JSON output with the data to connect to the SIP streaming. Maybe we could use this function to implement this proxy, however not sure if that would lock the device for other notifications.

**Comment by anthonyangel** [comment]
Thanks, I'd had a look at that already for 1-off uses it works well. The initial issues I can foresee are:
- Timeout on the live stream, I think that the URL needs to be refreshed every 10 minutes
- Whether having this connected locks out the device

**Comment by Klathmon** [comment]
@anthonyangel Do you have any information on how you made use of the SIP information? I'd like to see if I can look into this but i'm having a really tough time figuring out where to start.

**Comment by Klathmon** [comment]
I'm looking into this a bit more. I'm able to grab the video and audio using an SIP program (interestingly enough the stream only works without encryption... so i'm guessing that these videos are being sent across the net entirely un-encrypted...)

The stream returned from `live_streaming_json` expires in about 3 minutes for me. It seems that the `expires_in` field is in seconds, and it varies, requiring you to grab a new stream when it expires and maybe stitch them together at some point?

The doorbell still works while it's streaming, however i'm not sure if notifications will still go off on the phone when you are using it. I'll have to see if I can test this at another point, but I have a feeling it will work the same as I have been watching the stream on my phone (through the official app) while my wife tests it and it rang her phone, so I'm assuming it will work correctly.

I'm not the most versed in python, but there does seem to be a library for SIP in python that could be possibly used [here](https://github.com/AGProjects/python-sipsimple). With that it might be possible to grab a raw video stream out.

So TL;DR: it does look like this is possible, but it's not going to be all that easy, and it's probably way outside my ability in python.

**Comment by jlippold** [comment]
@Klathmon What SIP client did you use? There's are some sip libraries that support WebRTC with websockets but I'm shooting in the dark because I can't even get a simple player going.

**Comment by Klathmon** [comment]
I used the Blink client to do my testing. I was also looking at the WebRTC players that interface with SIP but decided to hold off as they wouldn't integrate as well with a python project like this (or home assistant).

By calling the SIP line I was able to get 2-way audio working instantly, but I had to go into Blink's Preferences and under "Accounts", switch the tab to "Media" and deselect "Encrypt audio and video" under "RTP Options" (i also unchecked "Send inband DTMF" but I don't think that applies here).

FYI if you have neighbors, the second you connect you will be broadcasting your computer's microphone to the ring speakers until you hang up! I think I scared the hell out of mine with a very loud "Holy shit it works!" the first time it connected!

**Comment by tchellomello** [comment]
That is very cool. Would be awesome to have this working in home assistant.

**Comment by jlippold** [comment]
That blink SIP client is expensive, let me know if you get video working, if you do then it's worth looking into an open source python lib

**Comment by Klathmon** [comment]
@jlippold I'm using the free/trial version and it is working well enough for now. It was also the first one I tried so others might work.

And I forgot to finish that sentence, it works with video if you uncheck using encryption.

**Comment by jlippold** [comment]
I assume you're on windows maybe? the mac app isn't free

**Comment by Klathmon** [comment]
Yes, windows.

It does seem to be bone-stock SIP so any client that let's you disable encryption *should* work for both audio and video.

**Comment by jlippold** [comment]
yea man, this works well.. For anyone wondering the sip address, is the `sip_to` property on the json. No other config needed, just that address. The SIP address expires after 10 seconds or so

**Comment by haimiko** [comment]
In the sip_to, is the is username-password@sipurl:port ?

**Comment by jlippold** [comment]
There's no password needed, just that `sip_to` property value... The "security" is that the sip to address expires frequently. FWIW, I only got it working with the windows version of that blink software/

**Comment by haimiko** [comment]
Guess I'll have to use a windows client.  I tried using an Android client
but i have yet to find one that will accept a passwordless SIP config.
 My goal would  be to have a python client continuously monitoring the sip
stream for motion and dings and recording as needed, so I can get off their
service.

**Comment by jlippold** [comment]
if you can dynamically pull the sip recording, let us know. I spent a whole day trying and gave up

**Comment by Klathmon** [comment]
You should be able to use any client that allows "Direct SIP dialing".

So when you setup a new SIP app, it will ask you for server, username, password, etc... You don't need any of that. You just need to be able to dial an SIP URI directly.

Other than that, you need to make sure you are using TCP not UDP (might be a setting in the options), and you need to ensure that you are not using any encryption (might be under a setting labeled RTP or something similar)

I just tried it with an android app MizuDroid and got it to work once, but you only have like 10 seconds from when you get the address till when you can't dial it any more.

**Comment by Klathmon** [comment]
Okay, i'm finally beginning to get this figured out.

So I've been doing some reading up on SIP ([RFC 3261](http://www.rfcreader.com/#rfc3261) and [RFC 6665](http://www.rfcreader.com/#rfc6665) are the big ones used).

I also managed to capture a full session using wireshark while i was testing with the Blink SIP client.

A few things i've found out:

SIP is just an "initiation" protocol, all it does is setup another protocol which actually streams the data. that protocol seems to be RTP (which from a preliminary look, should be easy to convert to something like an MP4 or something else which we can easily use output from). The SIP messaging responds with information on how to access this RTP stream, and it seems like it's randomized somewhat so we can't just skip the whole SIP messaging step and grab the stream.

It also looks like the version of SIP here uses some extensions (the second RFC I linked) to "subscribe" to something. That subscription request is being denied which is why i'm guessing these streams always end like 30 seconds after I start them.

I'm fairly sure we can get the SUBSCRIBE stuff to work by just impersonating the ring client, which means we can get a full stream (including 2-way audio if you want it) working.

All that being said, I'm gonna give this a shot. I haven't worked with python for like 7 years, so i'm hoping if I can get a "proof-of-concept" working, someone else can take the reins and actually get it working well.


My next step is to get some of this SIP messaging working in python. I'm going to use the Twisted library which does seem to have some support for SIP ([here's the protocol definition in twisted](https://github.com/twisted/twisted/blob/twisted-17.1.0/src/twisted/protocols/sip.py)).

**Comment by haimiko** [comment]
Awesome news. Will have a go.  Keep us posted on your progress.

**Comment by Klathmon** [comment]
Sorry to say but I think i'm throwing in the towel here (at least for a while).

I just don't know python well enough to be able to get anywhere, and i really don't think it will be very beneficial to this project if I spent the time to implement it in another language...

Here's some stuff I found, hopefully it's useful for someone to pick up:

* [p2p-sip](https://github.com/ikatson/p2p-sip) seems to have a pure-python sip implementation in it. at the very least it's inspiration.
* [SIPPing](https://github.com/pbertera/SIPPing) a SIP packet forging tool. This was what I got the furthest with, but still can't get a full send->response working even for the initial "INVITE" packet.
* [Twisted's SIP class](https://github.com/twisted/twisted/blob/twisted-17.1.0/src/twisted/protocols/sip.py). It does seem like this will work great, but I spent like an hour trying to figure out how to send the request before giving up.
* [Blink](http://icanblink.com/) the SIP client i've been using to do all of the exploration and testing. As long as you disable encryption in the settings as described further up in the thread, it works great. Make sure to make the request within like 30 seconds of calling `live_streaming_json`

Also, wireshark has proven to be super helpful in tracking down what is being requested and the responses.

I might pick this up in a week or so, but as of right now i'm just spinning and not getting anywhere.

**Comment by tchellomello** [comment]
@Klathmon  you helped a lot already. Thank you so much for studying the protocol. 

I want to get back to this RFE too which will be awesome to have this support. 

Thanks everyone involved and let's try to get this working. I'll try to look at the libraries you pointed out too.

**Comment by Klathmon** [comment]
If you or anyone else are able to get the python plumbing started, I can muddle my way through the actual implementation of the back-and-forth and grabbing the stream data. 

I actually love reverse engineering like that, it's just trying to do this in a language I haven't used for years is biting off more than I can chew.

**Comment by IanMitchell77** [comment]
I can get the alert and decode the JSON string but what did you put in the Blink SIP client to get the video? sip_to?

Many thanks in advance!

**Comment by Klathmon** [comment]
I believe it's the sip_to field (going from memory here). You do need to make sure you do this to get video working in the Blink SIP client:

>go into Blink's Preferences and under "Accounts", switch the tab to "Media" and deselect "Encrypt audio and video" under "RTP Options"

**Comment by IanMitchell77** [comment]
Thanks!

Did you find a Linux cli or gui that worked?

**Comment by IanMitchell77** [comment]
I can get AV with the Blink software!

However I only get a few seconds before it disconnects - does anyone have any idea on what I need to do to make it last longer?

**Comment by jlippold** [comment]
you have to requery the endpoint everytime it expires to get a new sip address

**Comment by IanMitchell77** [comment]
Maybe it is that, I'll check - Thanks!

It's seems that the sip token in the JSON response  may be relevant here but I'm not experienced in SIP and don't have an idea what to do with it. 

It takes me about 5 seconds to put the sip_to string in blink,  I then call and video lasts for a few seconds. I thought it expires much later than that but has anyone  automated getting the sip_to string in to blink and getting a new string when it finishes?

**Comment by rpavez** [comment]
Would be awesome if we could use http://www.jssip.net/ as sip client.

**Comment by Klathmon** [comment]
@rpavez  have you tried it with that yet? Just from taking a quick look there doesn't seem to be any reason why you can't use that!

There is a demo at https://tryit.jssip.net/ that you can try it with any SIP URI.

**Comment by rpavez** [comment]
I haven't I might try next weekend.

**Comment by jlippold** [comment]
I tried. I don't think ring uses websockets, and jssip does

**Comment by rpavez** [comment]
Another good option for fine grain control would be some command line sip client. 
[From Stackoverflow: SIP-Client for Raspberry Pi that works from command line?

](https://stackoverflow.com/questions/29435648/sip-client-for-raspberry-pi-that-works-from-command-line)

1. Linphone: Easy to install but I fail to make phone calls
RaspberryPI: Making SIP outbound calls using linphonec or an alternative SIP soft phone
2. Ring (formerly SFLphone): Looks promising but needs to be installed from source
3. PJSIP (C Library) See http://www.pjsip.org
4. Twinkle CLI

**Comment by IanMitchell77** [comment]
I tried the jssip website and that does not work.

Has anyone had any joy with the Blink software on Linux? I'm using Ubuntu 14.04 but can switch easily to make life easier, its only a VM.

**Comment by rpavez** [comment]
Found this article about building Blink from the source code for OSX https://gist.github.com/lucaspiller/8194862 maybe we can study how is implemented and extract necessary modules.

**Comment by rpavez** [comment]
Using wireshark and osx ring doorbell client I was able to capture the SIP call flow, it may help someone trying to implement a custom sip connecting figure out the process.
![screen shot 2017-06-12 at 12 27 37 am](https://user-images.githubusercontent.com/555910/27019372-0845cbe4-4f06-11e7-9edc-e01794464d89.png)

**Comment by cryptocake** [comment]
Has anyone tried to connect directly to the Ring doorbell? I port scanned mine and it showed 1 opened port: 1523. I'm trying to see if I can get any response from it RTP / SIP.

**Comment by IanMitchell77** [comment]
Any joy cryptocake?

**Comment by joeyberkovitz** [comment]
Artik cloud (https://artik.cloud/works-with/ring.html) gives access to the SIP URI, but indicates that there's no TLS on the one that it provides. Any chance we can connect through that?

**Comment by asantaga** [comment]
@Klathmon  can u give step by step instructions how to get the video streaming working from within Blink?

Ive followed what you have above and get no connection.. perhaps Im going mad.. or things have changed..

**Comment by asantaga** [comment]
As an interesting comment, I tried artik cloud (above from Joeyberkovitz) and whilst it did return a RTSP URL, the domainname wouldnt respond...  very strange..

**Comment by Sfinx** [comment]
To be able to get the video stream always you must send the custom INFO header just after ACK (connect) from their side. The INFO packet must look like:

```
....
INFO sip:.....
Via: SIP/2.0/TLS ....
From: <sip:....@ring.com>;tag=xxxx
To: "FS Doorbot" <sip:.....
CSeq: 21 INFO
Call-ID: ...
Max-Forwards: 70
Content-Length: 24
Content-Type: application/dtmf-relay
X-Ding: .... id from json...
X-Authorization: 
....

Signal=2
Duration=250
```

And just send BYE to close the SIP session to the doorbell

**Comment by jlippold** [comment]
It seems like we have all the details, but we dont have anyone who can write a SIP client in python

**Comment by Klathmon** [comment]
So I kind of abandoned this issue for a while, but I have done a bit more research..

I don't think Ring is going to let us stream from their servers for any extended amount of time. In my testing with some stuff "outside" of the things being explored in this issue, I found that even the actual apps from Ring will disable video if you try to keep it streaming for a few hours. Not to mention that they save all the video on their servers if you have the paid plan, and I'm almost 100% positive they will not be too happy and may start closing accounts if we fill their servers with tons of video data.

So while I think the SIP will work, we would probably want to limit it to the few minutes that the app allows, and go out of our way when programming our client to not abuse it and get the attention of Ring themselves.


That being said, if there is a way to directly talk to the ring device without touching their servers, that would be a much better solution. I'm going to be temporarily moving into a place soon where I won't be able to use the ring for a few months, so I am planning on wiring it up at my desk and doing a ton of packet-capture and seeing what it will accept/reject. Hopefully that will give us a better way of accessing this stuff that's much less likely to be shut down.

**Comment by jer78** [comment]
Personally I’d just like a feature that captures the last motion image and updates Home Assistant. If tapping on the image gets the live view, that would be cool too, or even just play the last motion recorded is fine for my purposes. 

My ultimate request would be for Home Assistant to send the video to my TV as a PIP video feed when there’s motion or a doorbell ring. But I think that’s a ways off. 

Thanks for everyone’s research into this. I’d happily code if I knew where to start but this is beyond my ability at the moment. 

Cheers!

**Comment by tchellomello** [comment]
@jer78  Please take a look at https://community.home-assistant.io/t/ring-doorbell/7943/237

I'm testing a prototype to playback the video on HASS. It would be awesome if you test it.

**Comment by jlippold** [comment]
@tchellomello your gist is for recorded videos, not live viewing correct?

**Comment by tchellomello** [comment]
@jlippold  correct!! We are not there yet for the live videos...

**Comment by untotren** [comment]
@klathmon live video api has been figured out already and documented:

http://www.kidder.io/2017/07/04/ring-doorbell-api/

**Comment by untotren** [comment]
@Klathmon were you able to make any progress on live video with the documented API?

http://www.kidder.io/2017/07/04/ring-doorbell-api/

**Comment by jplandry908** [comment]
Does anyone know if live_streaming_json still works? Every time I try to call it, it always returns 'None'. I have also tried triggering motion on the camera prior to calling it.

**Comment by Klathmon** [comment]
@untotren I haven't tried it yet. I'm in the process of moving and ended up accidentally shipping my ring to be held at storage at my new location with a lot of other stuff so I haven't had it to play with for a while.

Come April/May I may have more time to really dive into this once we are in the new house and settled in!

**Comment by egeste** [comment]
I just read through the doc. It looks like what's happening is:
1. The client application needs to send a request to the ring API to indicate that a user is awaiting a live video feed.
1. At some point within 17 seconds, the doorbell will poll the ring API to see if anyone requested a live video feed, thus returning a `200 OK` response.
1. Upon the `200 OK` response, the doorbell will perform a POST operation against the ring API to retrieve the necessary configuration to initiate an SIP connection. The response body of this request is noted as being unidentified by the author.
1. Upon receiving the necessary SIP configuration, the doorbell will connect to the SIP call and start streaming video.

It's interesting to note that the author of that article did not investigate the client-side behavior of the ring app. Further investigative work needs to be done to identify:

1. How do I initiate a request for a video on demand?
  * Can  be discovered by reverse engineering the app/sniffing traffic.
1. What kind of response does the API return once I've requested a video connection?
  * Probably returns some kind of encoded SIP connection information.
1. How do I connect to the video stream?
  * Again, probably SIP.

**Comment by GreenZapdos** [comment]
Has anyone been able to connect to the sip stream recently? Can you tell us what client you used to view the stream? It looks like the `doorbell.live_streaming_json` isn't working right now. `doorbell._ring.query(("https://api.ring.com/clients_api/dings/active?api_version=9&burst=1"), method='GET', raw=True)` works though and I was able to get a response from that. I put the value from `sip_to` into Linphone but wasn't able to get a connection. I'm not sure if I am using the client right though. There is a `sip_token` value that I think could be required to make the connection.

**Comment by HipsterZipster** [comment]
What's the current status on this issue? I'm interested in writing something for a Synology, javascript preferably

**Comment by jonFisher99** [comment]
I've avidly followed all the strategies discussed on here. It seems using the Artik Cloud to capture events and provide the SIP and RTSP URI's will be the optimal route to getting the live video feed. However, I'm finding A) The RTSP URI provided via Artik just doesn't work at all B) The SIP_TO parameter can be used with Blink SIP client, however only the audio NOT the video feed comes through, even if you remove the encryption (as suggested). Has anyone got a C# SIP client code example (perhaps using Ozeki?) to view the SIP_TO feed, or even just a SIP client software that will actually display the video feed?

**Comment by Doudy** [comment]
I've also avidly followed all the strategies discussed on here. 
But I am a little (much) lost.
What to do to see this famous video live?
What interests me is the live vision.
Is there a solution?
Thanks for your help.

**Comment by Niek** [comment]
Leaving this here, since the page itself is no longer available: https://web.archive.org/web/20180403003637/http://www.kidder.io/2017/07/04/ring-doorbell-api/

**Comment by kirantpatil** [comment]
Hi, I would like to know any opensource sip servers which is same as Ring servers.

**Comment by alekslyse** [comment]
Anyone got any further of connecting to the live stream. I cant see why it should be impossible as it is there - though even the windows app connects to 3.121.122.91 for auth. Just wish it would be a hidden rtsp stream or something running on the camera

**Comment by dgreif** [comment]
@rpavez @Sfinx I know I am really late to this thread, but I am working on implementing live video from Ring in my [homebridge-ring plugin](https://github.com/dgreif/ring/issues/35).  I am _very_ close to having it working, but can't quite crack it.  I am successfully negotiation the SIP setup and am receiving RTCP packets on the port that I give to Ring, but am not receiving any RTP packets for some reason.  What I really need at this point is a wireshark dump of the SIP and RTP dialog similar to the image that @rpavez posted a couple years ago.  Would either of you be willing to supply one, or better yet, walk me through how you managed to capture the live session from the app?  Feel free to respond over on https://github.com/dgreif/ring/issues/35. Thanks to everyone here who supplied info and links on this issue as it has gotten me 95% of the way there!

**Comment by cdoublejj** [comment]
well for wifi ring idk but, for PoE RING like the Elite I have wouldn't one mirror the port on the switch and collect it with wire shark? i'v never done it before my self. i have a managed switch and a VM running shark.

**Comment by cdoublejj** [comment]
> @rpavez @Sfinx I know I am really late to this thread, but I am working on implementing live video from Ring in my [homebridge-ring plugin](https://github.com/dgreif/ring/issues/35). I am _very_ close to having it working, but can't quite crack it. I am successfully negotiation the SIP setup and am receiving RTCP packets on the port that I give to Ring, but am not receiving any RTP packets for some reason. What I really need at this point is a wireshark dump of the SIP and RTP dialog similar to the image that @rpavez posted a couple years ago. Would either of you be willing to supply one, or better yet, walk me through how you managed to capture the live session from the app? Feel free to respond over on [dgreif/ring#35](https://github.com/dgreif/ring/issues/35). Thanks to everyone here who supplied info and links on this issue as it has gotten me 95% of the way there!

well for wifi ring idk but, for PoE RING like the Elite I have wouldn't one mirror the port on the switch and collect it with wire shark? i'v never done it before my self. i have a managed switch and a VM running shark.

**Comment by dgreif** [comment]
I managed to get the SIP messages using a capture tool on android.  I have live streaming video fully functional in my homebridge-ring plugin now.  Currently working on some refactoring to get audio working as well, but the basic SIP setup is solid at this point.  Once the call is established, Ring's servers will send H264 video and PCMU audio over UDP to the ip/ports you give them in the SIP invite.  Should be fairly portable to python if anyone has a good grasp of javascript and python.

**Comment by rpavez** [comment]
> @rpavez @Sfinx I know I am really late to this thread, but I am working on implementing live video from Ring in my [homebridge-ring plugin](https://github.com/dgreif/ring/issues/35). I am _very_ close to having it working, but can't quite crack it. I am successfully negotiation the SIP setup and am receiving RTCP packets on the port that I give to Ring, but am not receiving any RTP packets for some reason. What I really need at this point is a wireshark dump of the SIP and RTP dialog similar to the image that @rpavez posted a couple years ago. Would either of you be willing to supply one, or better yet, walk me through how you managed to capture the live session from the app? Feel free to respond over on [dgreif/ring#35](https://github.com/dgreif/ring/issues/35). Thanks to everyone here who supplied info and links on this issue as it has gotten me 95% of the way there!

I used https://ring.com/gettheapp got it connected to my ring device and then https://www.wireshark.org/download.html to check traffic filtered by SIP traffic only.

**Comment by cdoublejj** [comment]
So once it's captured i assume it would take some serious doing and CPU power to convert to ONVIF or RTSP or the likes for say sending over to Blue Iris?

**Comment by dgreif** [comment]
@cdoublejj as long as your target software supports h264, it won’t take much CPU at all. For HomeKit, I literally just pass along the video UDP packets without touching them. For audio, I have to convert from PCM to aac-eld because HomeKit is picky, but I do that with ffmpeg and it takes barely any CPU.

**Comment by jeroenterheerdt** [comment]
I made a very quick and dirty solution that basically wraps around the node solution made by @dgreif. Hopefully someone with more skills in Node can turn this into a proper Python solution. The PR is here if you want to try: https://github.com/tchellomello/python-ring-doorbell/pull/133

**Comment by cdoublejj** [comment]
that's awesome!


Blue Iris supports H264 AND H265, I think it also supports more than just ONVIF and RTSP, think it may support a few other or at the very least it appears to have a range of protocols and camera types. It is a pretty robust solution! Heck i'd consider a bounty $$$ if wasn't a lofty goal.

**Comment by jeroenterheerdt** [comment]
I am not interested in Blue Iris :) I just want this to work in Home Assistant :)

**Comment by Devqon** [comment]
Is someone still on this in python? I was working further on the node solution by @jeroenterheerdt in a HA addon, but then thought why can't this just be done in the language of HA itself (python). I am not very much of a python developer myself, but willing to try to port the node solution to python.

**Comment by dgreif** [comment]
If anyone does decide to take this on in Python, I've documented a lot of my findings from implementing it in node: https://github.com/dgreif/ring/wiki/SIP-Communication.

**Comment by grtessman** [comment]
Hey guys... I skimmed through this thread a bit.  I noticed that someone said SIP streaming URL is returned.  Why would you need any client for that? All modern browsers support WebRTC using SIP in a native browser VIDEO element.  The real question is, does this still work without a subscription today for the live stream?  Seems like it should...

**Comment by Shamshala** [comment]
LiveView proxy could be really cool feature! I would be twice as glad now because i've just changed ISP and it uses TDD Flexible Framing which corrupts my streams (sound primarily). Quite pity with 4 four Ring cameras. :unamused:

**Comment by Dfbrand** [comment]
Any update on this at all? I am not confident enough in python unfortunately

**Comment by kethan1** [comment]
Look at this: http://cs.uccs.edu/~cs526/svoip/src/shtoom/shtoom/sip.py, it is messy, and I can't make heads or tails out of it :)

**Comment by shadtorrie** [comment]
Has there been progressed made on this issue? I would very much like to have access to a live feed. I have some python experience and would love to help out here.

**Comment by levidavis** [comment]
same here!

**Comment by goodboy** [comment]
Not to be involved too much (mostly snooping for a friend) but y'all could use [SIPp](https://github.com/SIPp) probably with a script [similar to this](https://tomeko.net/other/sipp/scenarios/INVITE_SDP_video.xml) (but probably ideally with [embedded SDP parsing for video](https://github.com/SIPp/sipp/blob/master/sipp_scenarios/pfca_uas_video_crypto_simple.xml)) to pull out the RSTP stream deatz.

Once you have the media deats it should be as simple as connecting a media client as per:
- https://unix.stackexchange.com/questions/25368/how-to-play-rtsp-stream-under-linux-for-the-bosch-autodome-junior-hd-iva-camera
- https://stackoverflow.com/questions/71824807/play-rtsp-stream-from-webcam-using-gstreamer

If you really wanted to get fancy with python you probably want to use [bindings to pjsip](https://trac.pjsip.org/repos/); pure python solutions for SIP [died not long ago](https://github.com/Eyepea/aiosip/issues/136#issuecomment-748930217).

Anyway hope this helps those of you still hanging on this.

**Comment by dgreif** [comment]
If it helps, Ring is now using WebRTC for streaming with WebSockets as the messaging channel.  Not sure if those have better support in python

**Comment by goodboy** [comment]
@dgreif i presume you sniffed such things in the dev-mode of a browser orr?

Any hot tips are highly appreciated 🏄🏼

**Comment by dgreif** [comment]
Yep, I've got a full implementation in https://github.com/dgreif/ring/tree/master/api/streaming.  It functions slightly different between normal cameras and cameras connected via Ring Edge, but overall it's the same setup with WebSockets and WebRTC
